lib/Conversion/AIEVecToLLVM/AIEVecToLLVM.cpp

```diff
@@ -921,76 +912,48 @@ class MatMulOpConversion
     decodedMatMulOp.acc = rewriter.create<vector::ShapeCastOp>(
         loc, accFlattenedVecTy, decodedMatMulOp.acc);
 
-    if (decodedMatMulOp.kind != DecodedMatMulOp::Kind::BF16) {
-      if (lhsFlattenedVecTy.getShape()[0] *
-              lhsFlattenedVecTy.getElementTypeBitWidth() !=
-          512)
-        decodedMatMulOp.lhs =
-            widenVectorTo512bit(rewriter, loc, decodedMatMulOp.lhs);
-      decodedMatMulOp.lhs =
-          rewriter
-              .create<LLVM::BitcastOp>(
-                  loc, VectorType::get({64}, rewriter.getI8Type()),
-                  decodedMatMulOp.lhs)
-              .getResult();
-      if (rhsFlattenedVecTy.getShape()[0] *
-              rhsFlattenedVecTy.getElementTypeBitWidth() !=
-          512)
-        decodedMatMulOp.rhs =
-            widenVectorTo512bit(rewriter, loc, decodedMatMulOp.rhs);
-      decodedMatMulOp.rhs =
-          rewriter
-              .create<LLVM::BitcastOp>(
-                  loc, VectorType::get({16}, rewriter.getI32Type()),
-                  decodedMatMulOp.rhs)
-              .getResult();
-      decodedMatMulOp.acc =
-          rewriter
-              .create<LLVM::BitcastOp>(
-                  loc, VectorType::get({16}, rewriter.getI64Type()),
-                  decodedMatMulOp.acc)
-              .getResult();
-    } else
-      decodedMatMulOp.acc =
+    Type i32ty = rewriter.getI32Type();
+    auto confCst = rewriter.create<LLVM::ConstantOp>(
+        loc, i32ty, rewriter.getI32IntegerAttr(decodedMatMulOp.conf));
+    SmallVector<Value> operands({decodedMatMulOp.lhs, decodedMatMulOp.rhs,
+                                 decodedMatMulOp.acc, confCst});
+    Value matMulResVal;
+    if (decodedMatMulOp.kind == DecodedMatMulOp::Kind::BF16)
+      matMulResVal =
           rewriter
-              .create<LLVM::BitcastOp>(
+              .create<aievec::MacConfBF16IntrOp>(
                   loc, VectorType::get({8}, rewriter.getI64Type()),
-                  decodedMatMulOp.acc)
+                  forceCastOperandsToSignature(
+                      rewriter, loc, operands,
+                      {VectorType::get({32}, rewriter.getBF16Type()),
+                       VectorType::get({32}, rewriter.getBF16Type()),
+                       VectorType::get({8}, rewriter.getI64Type()), i32ty}))
```

Comment:
Is it possible to get the signature from the intrinsic op's definition?

Reply:
Not out of the box, no. The plan was to define an extra function that returns the signature, but it's not straightforward. I'll give it a go in the future.